### PR TITLE
Enable Count(*) syntax

### PR DESCRIPTION
**Pull request by jeffreylovitz**
Enable COUNT(*) projections, emit validation error on `COUNT(DISTINCT *)` projections.

### CODE REVIEW

File: src/ast/ast.c [hunk 1]
```diff
@@ -25,11 +25,17 @@ static inline void _prepareIterateAll(rax *map, raxIterator *iter) {
...
+	bool apply_all = (cypher_astnode_type(expression) == CYPHER_AST_APPLY_ALL_OPERATOR);
+
+	// Retrieve the function name and add to rax.
+	const cypher_astnode_t *func = (!apply_all) ? cypher_ast_apply_operator_get_func_name(expression) :
+								   cypher_ast_apply_all_operator_get_func_name(expression);
 	const char *func_name = cypher_ast_function_name_get_value(func);
 	raxInsert(referred_funcs, (unsigned char *)func_name, strlen(func_name), NULL, NULL);
 
+	if(apply_all) return;  // Apply All operators have no arguments.
+
```
Comment (swilly22) [inline src/ast/ast.c]: In which case `uint narguments = cypher_ast_apply_operator_narguments(expression);`
`narguments` should be 0 and we'll simply skip the following loop,

If that's indeed the case I prefer we remove this line and simply fall throw, one logic which works for all cases.

File: src/ast/ast_build_ar_exp.c [hunk 2]
```diff
@@ -88,6 +88,29 @@ static AR_ExpNode *_AR_EXP_FromApplyExpression(const cypher_astnode_t *expr) {
...
+	const cypher_astnode_t *func_node = cypher_ast_apply_all_operator_get_func_name(expr);
+	const char *func_name = cypher_ast_function_name_get_value(func_node);
+
+	bool distinct = cypher_ast_apply_all_operator_get_distinct(expr);
+	if(distinct) {
+		// TODO consider opening a parser error, this construction is invalid in Neo's parser.
+		char *error;
+		asprintf(&error, "Cannot specify both DISTINCT and * in the function call %s", func_name);
+		QueryCtx_SetError(error); // Set the query-level error.
+	}
```
Comment (swilly22) [inline src/ast/ast_build_ar_exp.c]: Should we catch this in the AST validation phase?

File: src/ast/ast_build_ar_exp.c [hunk 3]
```diff
@@ -88,6 +88,29 @@ static AR_ExpNode *_AR_EXP_FromApplyExpression(const cypher_astnode_t *expr) {
...
+
+	op = AR_EXP_NewOpNode(func_name, 1);
+
+	// Introduce a fake child constant so that the function always operates on something.
+	op->op.children[0] = AR_EXP_NewConstOperandNode(SI_BoolVal(1));
```
Comment (swilly22) [inline src/ast/ast_build_ar_exp.c]: I believe is `*` only applicable for count aggregations, it doesn't make sense to do `return avg(*)` we should add an AST validation for that.

File: src/ast/ast.c (hunk 1 above)
Comment (jeffreylovitz) [reply src/ast/ast.c]: Unfortunately, Apply All operators are parser objects of a different type and with a different interface than Apply operators (thus the conditional logic for name retrieval at https://github.com/RedisGraph/RedisGraph/pull/717/files#diff-51a289a0066db87d439f3b280ed18df4R32-R33).

This is to say that `cypher_ast_apply_operator_narguments` will cause a type failure when given an Apply All operator.

File: src/ast/ast_validations.c [hunk 4]
```diff
@@ -207,6 +207,69 @@ static AST_Validation _ValidateReferredFunctions(rax *referred_functions, char *
...
+	for(uint i = 0; i < child_count; i ++) {
+		const cypher_astnode_t *child = cypher_astnode_get_child(node, i);
+		AST_Validation res = _VisitFunctions(child, func_names, reason);
+		if(res != AST_VALID) return res;
+	}
+
+	return AST_VALID;
+}
+
+static AST_Validation _ValidateFunctionCalls(const cypher_astnode_t *node, char **reason,
```
Comment (swilly22) [inline src/ast/ast_validations.c]: Shouldn't there be a call to `_ValidateFunctionCalls` from `_Validate_WITH_Clauses` and `_Validate_CREATE_Clauses`
`WITH none_existing_func() AS x, count(distinct *) AS z`
`CREATE ({x:none_existing_func(), z:count(distinct *)}) `

**Comment by swilly22** [comment]
I'll merge this but let's introduce the additional validations.

File: src/ast/ast_validations.c (hunk 4 above)
Comment (jeffreylovitz) [reply src/ast/ast_validations.c]: :+1: Good call, we should have had similar calls all along.

**Pull request merged**